### PR TITLE
[WIP] Move drop elaboration before borrowck

### PR DESCRIPTION
**Pull request by zeegomo**
Proof of concept to show what's needed to move drop elaboration before borrow checking. Ideally we could have it right after MIR build, but this tries to be as minimal as possible.

There are a few hacks, mostly around drop-elaboration code that does not pass borrowck / move analysis.

I plan to work on the remaining failing tests in the coming days

### CODE REVIEW

**Comment by WaffleLapkin** [comment]
r? compiler

**Comment by Zeegomo** [comment]
My bad, I tested the last couple refactorings without a complete build, will fix it later today

File: compiler/rustc_mir_transform/src/lib.rs [hunk 1]
```diff
@@ -509,7 +511,6 @@ fn run_runtime_lowering_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
     let passes: &[&dyn MirPass<'tcx>] = &[
         // These next passes must be executed together
         &add_call_guards::CriticalCallEdges,
-        &elaborate_drops::ElaborateDrops,
```
Comment (tmiasko) [inline compiler/rustc_mir_transform/src/lib.rs:512]: The other pass should run just before the drop elaboration still. (It breaks critical edges and makes it possible to execute extra code when a function call returns by inserting code to a destination block.)

File: compiler/rustc_mir_transform/src/lib.rs [hunk 2]
```diff
@@ -339,6 +339,10 @@ fn mir_promoted(
         &[
             &promote_pass,
             &simplify::SimplifyCfg::new("promote-consts"),
+            // These next passes must be executed together
+            &add_call_guards::CriticalCallEdges,
```
Comment (bjorn3) [inline compiler/rustc_mir_transform/src/lib.rs:343]: Doing `CriticalCallEdges` before optimizations means that all optimizations now have to maintain the invariant that the destination edge of a call is not a critical edge. `SimplifyCfg` doesn't maintain this invariant afaik. Violating this invariant will result in codegen backends incorrectly handing return value storage for functions that can unwind.

**Comment by bjorn3** [comment]
It looks like this will conflict with https://github.com/rust-lang/rust/pull/106430. This PR removes dead unwinds before drop elaboration, but borrowck needs to know about these dead unwinds.

**Comment by Zeegomo** [comment]
> but borrowck needs to know about these dead unwinds.

Could you please elaborate on why?
I was under the impression that borrowck cared about the special `False{Edge, Unwind}` (which are untouched by both this and https://github.com/rust-lang/rust/pull/106430) but not 'normal' unwinds more than any other dead block (which are already removed before borrowck by a few passes of `SimplifyCfg`)

**Comment by bjorn3** [comment]
Borrowck cares about all jump and unwind edges. FalseEdge and FalseUnwind are used for when during mir building it is already known that the jump and unwind edges can't happen and there is nothing to attach those edges to otherwise. For example in case of an infinite loop. Drop elaboration would need to add FalseUnwind to preserve unwind edges for borrowck if it is moved before borrowck. As for why borrowck cares, see the doc comment on FalseUnwind:

> A terminator for blocks that only take one path in reality, but where we reserve the right to unwind in borrowck, even if it won’t happen in practice.
> [...]
> The imaginary cleanup block link. This particular path will never be taken in practice, but in order to avoid fragility we want to always consider it in borrowck. We don’t want to accept programs which pass borrowck only when panic=abort or some assertions are disabled due to release vs. debug mode builds. This needs to be an Option because of the remove_noop_landing_pads and abort_unwinding_calls passes.

**Comment by Dylan-DPC** [comment]
@Zeegomo any updates on this?

**Comment by Zeegomo** [comment]
Unfortunately I don't have the bandwidth at the moment to continue this work

**Comment by Dylan-DPC** [comment]
Closing this as inactive. Feel free to reöpen this pr or create a new pr if you get the time to work on this. Thanks

**Pull request closed without merging**